### PR TITLE
Added setting of minimum and maximum curves values on the graphs. VITE Web version. 

### PR DESCRIPTION
**Pull request by demvlad**
PR #689 for VITE Web.

### CODE REVIEW

**Comment by nerdCopter** [comment]
maybe not be properly rebased.
image
& Conflicting files

**Comment by demvlad** [comment]
> maybe not be properly rebased. 

Maybe!
I do not understand. I've made rebase and resolved all conflicts.
But there were too much conflicts after pushing branch. I've resolved their in the github online interface.
Maybe this version will work:)
If it will not, then i will repeate rebase.

**Comment by haslinghuis** [comment]
Probably need to drop only these ones:

image

**Comment by demvlad** [comment]
I've tryed to load this version from github and run it.
This is very wunderfull:), but it seems, the program is working!
Will study, how to do rebase properly...

**Comment by haslinghuis** [comment]
`git rebase -i HEAD~153` and drop the 3 commits. Save and force push.

**Comment by nerdCopter** [comment]
i like `git rebase -i` (interactive) as well. replaceing `pick` with `drop` for the suspect commits.
when i perform that exact command, my editor shows 4 to maybe drop.
image

**Comment by demvlad** [comment]
Thank's.
Can i do rebase like:
rebase  -i   my_min_max_branch  vite_veb_branch
or
rebase -i  vite_veb_branch my_min_max_branch

What rebase direction is better?

**Comment by haslinghuis** [comment]
Something like that described here: https://git-scm.com/docs/git-rebase

**Comment by demvlad** [comment]
> looks good 🚀 

I've done rebase more accuratly.
The test master merge has not any conflicts.
I am going to push this version after a small bit time.

**Comment by nerdCopter** [comment]
i wonder now that it is PWA, are built-in browser based menus still bugged?

**Comment by demvlad** [comment]
> i wonder now that it is PWA, are built-in browser based menus still bugged?

The olde nw.js menu?

**Comment by demvlad** [comment]
> > looks good 🚀
> 
> I've done rebase more accuratly. The test master merge has not any conflicts. I am going to push this version after a small bit time.

Yes! The all rebase for wite web have not merge conflicts.
... but other not web branches have some conflicts:) Will resolve...

**Comment by demvlad** [comment]
Improvements:
The menu is simple now. There are no submenues by default.
menu

If the stanard menu is not enough for you and you want more features, then click menu item while the Shift key is pressed. 
And you will see the submenu with extended actions.

The Save item is removed from the menu, because it duplicated the 'Save' checkboxes in Chart setup dialog box.
You need to use the 'Save' checkboxes in one case only: if you want to change default min-max values for some curves by using workspace slots or by loading next logs.
If 'Save' checkbox is not checked for curve then it will use default presets min-max values by loading next logs or workspace charts preset.
I do not like the 'Save' checkboxes name, because the charts configuration will save in any case when you press the Apply button at dialog box or press Shift+num key after that.
I want to find other short name for these checkboxes.

**Comment by demvlad** [comment]
It seems, there are all till.
All are ready for testing.

**Comment by demvlad** [comment]
The workspace save strategy is changed.

I think, that are all till.
I am inviting all participants for testing.

**Comment by demvlad** [comment]
The 'This curves' actions replaced into submenu.
Finaly we have next  variants menu:
### Main menu default


### Main menu with active submenu while the Shift key is pressed


### 'This curves' actions submenu


### Main menu if the chart has one curve only

**Comment by demvlad** [comment]
The optimization process is endless :)
But i think there are all till....

**Comment by demvlad** [comment]
**Improvements:**
The Apply and Cancel buttons are enabled while menu is open. You can use it to close the dialog box without pressing of Back and Return menu items.
The Alt (optionKey on Mac) key is used instead of Shift to use hidden submenu.
The 'Full range' item of 'This curve' menu has submenu while Alt key is pressed.
The 'Default' item on main menu has hidden submenu too.
The menu source code structure is improved.

The issue of this version what i know, and their resolving at moment:
- sometime the Esc key is not working - to close menu by click at Back and Return menu item.
- sometime the right arrows are not showed on menu items while Alt key is pressed - continue to use submenu with Alt-key. The submenu are working really.
There issues appear sometime after some mouse or keyboard actions while the menu is openned. There are some keyboard event handlers problem, i will resolve these.

**The 'Full range' submenu on 'This curve' menu while Alt key is pressed.**
sub3


sub_single

File: src/graph_minmax_setting_menu.js [hunk 1]
```diff
@@ -547,21 +628,23 @@ export function showMinMaxSetupContextMenu(menu_pos_x, menu_pos_y, selected_fiel
     }
 
     function SetZoomToCurves (e) {
-
-        if(e.shiftKey == false) {
+        const target = $(e.target);
+        const SingleCurve = target.hasClass('SingleCurve');
+        if(e.altKey == false || SingleCurve) {
```
Comment (haslinghuis) [inline src/graph_minmax_setting_menu.js]: :space_invader:

File: src/graph_minmax_setting_menu.js [hunk 2]
```diff
@@ -701,90 +777,101 @@ export function showMinMaxSetupContextMenu(menu_pos_x, menu_pos_y, selected_fiel
...
+        menu.append(elem);
+
+        elem = $('<div class="ZoomOut SingleCurve">Zoom Out&nbsp;&nbsp;&nbsp;&nbsp;</div>');
+        elem.click(SetZoomToCurves);
+        menu.append(elem);
+
+        if (is_main_menu)
+            elem = $('<div class="iconDiv SingleCurve">Full range<span class="right-arrow" style="display: none">&#9658;</span></div>');
+        else
+            elem = $('<div class="iconDiv SingleCurve">Full range<span class="right-arrow2" style="display: none">&#9658;</span></div>');
```
Comment (haslinghuis) [inline src/graph_minmax_setting_menu.js:821]:
```suggestion
        elem = $(`<div class="iconDiv SingleCurve">Full range<span class=${is_main_menu ? "right-arrow" : "right-arrow2"} style="display: none">&#9658;</span></div>`);
```

**Comment by demvlad** [comment]
I've found that in Vite Web version the Alt-key activate the browsers menu in Windows. Therefore in the Vite Web version i am using Shift-key.

File: src/graph_minmax_setting_menu.js (hunk 2 above)
Comment (demvlad) [reply src/graph_minmax_setting_menu.js]: Thank's! Will know about these syntax.

**Review by demvlad — COMMENTED** [review]
Code style improvement

**Review by demvlad — COMMENTED** [review]
Code style improvement

**Review by demvlad — COMMENTED** [review]
Code style improvement

**Review by demvlad — COMMENTED** [review]
Code style improvement

**Comment by demvlad** [comment]
@haslinghuis I am seeng more yours code style improvements in several files.
But i do not know how to apply it.
Please, could you help me.

**Comment by demvlad** [comment]
@ctzsnooze Please, check this update.
1. I've disable input autocompletion
2. The active top level 'Return' or 'Back' menu items have more smart design

**Comment by demvlad** [comment]
@ctzsnooze 
I've revert the all autocomplete preventions commits. Please check this version.

**Comment by demvlad** [comment]
Improvements:
- the curves expo is set to 100% as default (i did not remove it full till).
- added menu items actions to set min-max for marker selected time interval in extended submenu
-

**Comment by demvlad** [comment]
Updated:
- the 'Keep' checkboxes are removed from Charts setup dialog box
- the 'Grid' checkboxes are removed from Charts setup dialog box, the grid visibility is control by On/Off button
- the main menu items actions close the menu

**Comment by demvlad** [comment]
Resolved error of data storage by using Expo, Grid, Smooth buttons

**Comment by demvlad** [comment]
**THE MANUAL**
The MinMax control PR can control curves scale and placement quickly and full.
The current MinMax curves settings are showed at 'Configure graphs' dialog box in the "Minimum" and "Maximum" columns.


The MinMax values can be changed:
- By direct input into table cells at 'Configure graphs' dialog box. It is possible to changes values manualy or set default values by double mouse click at values field.
- By using context menu

To show context menu you must do right mouse click on Minimum or Maximum values field what you want to edit.
The main context menu:


The main menu include following parts:
1. The menu actions to edit all curves
2. The menu actions to edit one curves what you selected by right mouse click (The 'Gyro pitch' for example at the picture)


The actions for all curves:
- 'Like this one' - apply MinMax of curve what you selected by right mouse click ('Gyro pitch' for example) to all curves.
- 'All to default' - set all curves MinMax values to default
- 'Zoom in', 'Zoom out' - change zoom for all curves
- 'Default' - set the all curves MinMax values to default
- 'Full range' - set the all curves MinMax values from Log file data during the all flight time
- 'One scale' - set all curves to one scale
- 'Centered' - The curves are centered around X-axis

The single curve submenu has same actions: 


The main menu has extended mode. You need press the 'Shift' key to activate it and select the submenu, what you need. 


**The 'Like this one' extended submenu.**
You can change MinMax values and select curves what you need by using the checkboxes to apply these values. 
Click 'SET MIN-MAX VALUES' item for apply. 
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The 'Zoom in', 'Zoom out' extended submenu.**
You can set the zoom procent value and select curves what you need by using the checkboxes to apply zoom. 
Click 'ZOOM IN', 'ZOOM OUT' items for apply zoom
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The 'Default' extended submenu.**
You can select curves what you need by using the checkboxes to apply default values. 
Click 'SET CURVES TO DEFAULT' item for apply
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The 'Full range' extended submenu.**
You can select curves what you need by using the checkboxes to apply values. 
Click 'At all global log time' menu item to set MinMax values from log data during all time.
Click 'At local window time' menu item to set MinMax values from current time interval at the chart window.
Click 'At markere time range' menu item to set MinMax values from markered time interval what you select by using "I", "O" keys. If it is not select then will apply all log time interval.
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The 'One scale' extended submenu.**
You can select curves what you need by using the checkboxes to set the same scale. 
Click 'SET SET CURVES TO SAME SCALE' item for apply
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The 'Centered' extended submenu.**
You can select curves what you need by using the checkboxes to centered. 
Click 'SET CURVES TO ZERO OFFSET' item for apply
Click 'Back' menu item to go back to main menu. 
Click 'Apply change' or 'Cancel' button on the main 'Configure graphs' dialog box to close the menu and dialog box immediately


**The single curve submenu has one 'Full range' extended item:** 
It is like 'Full range' extended submenu, but for one selected curve only.


If you open context menu for one curves chart, then you see the short menu:

**Comment by ctzsnooze** [comment]
Working REALLY WELL!!  Absolutely love this PR.
I can now quickly and easily set up the display exactly how I need it.
Particularly helpful for debugs that don't have sensible scaling info.
Ready to merge?  I think so.
Definitely ready for review.

**Review by SupaflyFPV — APPROVED** [review]
Handy feature I like it...good for debugs and cough analysis of other firmwares...

**Review by haslinghuis — DISMISSED** [review]
Awaiting more testing 
- as the PR is huge 
- it's impossible to review
 - as having reviewed before @demvlad has been educated :smiley:

Code rules should be automated @chmelevskij - agree we should not have to bother with code styling

**Comment by ctzsnooze** [comment]
May require a new workspaces file.
Here's my current file.
[workspaces-ctzsnooze.json.zip](https://github.com/betaflight/blackbox-log-viewer/files/15135959/workspaces-ctzsnooze.json.zip)

**Comment by demvlad** [comment]
@haslinghuis 
Yes, we have two big improvements at moment: the VITE WEB and my MinMax.
The all are need testing.
Therefore i will continue to support my non Web PR version to different some bugs, what are possible.

**Comment by demvlad** [comment]
@ctzsnooze workspace is used by default.
His workspace uses curves min-max values, what this PR can control.

**Comment by haslinghuis** [comment]
:thinking: we should add default workspace in separate PR as there is some discussion about having multiple `presets` as default.

Anything that could reduce the size of this PR :stuck_out_tongue: 

See also: #726

**Comment by demvlad** [comment]
> 🤔 we should add default workspace in separate PR as there is some discussion about having multiple `presets` as default.

These are a small different workspaces. This workspace PR stored any min-max values what you can set to scale and place curves at the chart. 
If the full web version can not save local data and need some multy-presets, probably i will prefer developers local variant :)

**Comment by demvlad** [comment]
The small improvement:
The mouse click at curves on the legend works like toggle button (make the analyser chart as visible/unvisible)

**Comment by ctzsnooze** [comment]
@haslinghuis @nerdCopter @SupaflyFPV 
This PR is ready to merge now

**Review by SupaflyFPV — APPROVED** [review]

**Review by haslinghuis — APPROVED** [review]

**Comment by nerdCopter** [comment]
i feel like `Default` should be first in menu -- thoughts?
image

**Comment by demvlad** [comment]
> i feel like `Default` should be first in menu -- thoughts? 

Yes, maybe.

**Comment by demvlad** [comment]
Updated:
- The all charts configuration changes are previewed now
- The menu items order is changed: the Zoom and Default items were moved to bottom of menu as less helpfull
- some issue was resolved

File: src/graph_legend.js [hunk 3]
```diff
@@ -72,23 +72,21 @@ export function GraphLegend(targetElem, config, onVisibilityChange, onNewSelecti
...
+					config.selectedGraphIndex    = selectedGraphIndex;
+					config.selectedFieldIndex    = selectedFieldIndex;
+					if (onNewSelectionChange) {
+						onNewSelectionChange();
+					}
+				}
+				else {
+					onNewSelectionChange(true);
+				}
+            }
```
Comment (haslinghuis) [inline src/graph_legend.js:88]: - tabs need to be spaces :) 
- else clause should follow on same line as closing bracket

```suggestion
            if (!e.altKey) {
                const selectedFieldName = config.getGraphs()[selectedGraphIndex].fields[selectedFieldIndex].friendlyName;
                if (config.selectedFieldName != selectedFieldName) {
                    config.selectedFieldName     = selectedFieldName;
                    config.selectedGraphIndex    = selectedGraphIndex;
                    config.selectedFieldIndex    = selectedFieldIndex;
                   if (onNewSelectionChange) {
                       onNewSelectionChange();
                    }
                } else {
                   onNewSelectionChange(true);
                }
            }
```

**Review by haslinghuis — APPROVED** [review]

**Comment by nerdCopter** [comment]
The `Sonar Cloud` shows that many switch/case just return the same thing `return value;` so it seems the switch/case are not even needed. please verify or reply otherwise.

**Comment by demvlad** [comment]
> The `Sonar Cloud` shows that many switch/case just return the same thing `return value;` so it seems the switch/case are not even needed. please verify or reply otherwise.

switch/case ??? Yes, there are in default curves and units converters code. I've used the previouse source code as the ground. There are very bigger switch/case code really. Ok, I will check it.

**Comment by demvlad** [comment]
@nerdCopter 
The example of code:
case 'FFT_TIME':
                    switch (fieldName) {
                        case 'debug[0]':
                            return value;
                        case 'debug[1]':
                            return value;
                        // debug 2 = not used
                        // debug 3 = not used
                        default:
                            return value;
                    }
The all cases return back the same input value today. But maybe someone will edit this or similar DEBUGs next time, i do not know that.

**Comment by haslinghuis** [comment]
Will need to be refactored in another PR some time.

**Review by nerdCopter — APPROVED** [review]

**Comment by nerdCopter** [comment]
okay, i approved. if refactoring for a future date, no problem.

**Review by ctzsnooze — APPROVED** [review]
Have used this for a long time now, let's get it merged.

I would prefer 'default' at the bottom of the list'.  Reason is that I want the 'most useful option' at the top, and 'default' is generally the least useful option :-)  The most useful option is 'all like this'; It's the shortcut to use after setting up the scaling on one trace when all should be the same. Default is only needed in the unlikely event that you are curious what the defaults used to be like.  Mostly the user will experiment with the scaling options and make them however they like.  Rather than go back to default, they will typically just change things to suit their preference.  Hence, at the bottom is a good place for the Default option.

Main issue with this PR is that the user's workspaces have to be rebuilt.

Once merged, we can fix the above case statement issues, check for suitable default min-max values for debug traces, and so on.

**Comment by demvlad** [comment]
> Main issue with this PR is that the user's workspaces have to be rebuilt.

@ctzsnooze 
It is possible to use olde workspaces in this PR. 
It has default curves MinMax values as start after loading.   
I've checked this on supafly.json file for example.

**Comment by nerdCopter** [comment]
Although useful, this file should not have been committed:
src/workspaces-ctzsnooze.json

**Comment by demvlad** [comment]
The link to manual:

https://github.com/betaflight/blackbox-log-viewer/pull/726#issuecomment-2079546532

**Comment by demvlad** [comment]
> Although useful, this file should not have been committed: src/workspaces-ctzsnooze.json

This is workspace as default, it uses in main.js

**Comment by haslinghuis** [comment]
@demvlad huge thank you for all improvements. Very much appreciated. Would love to see the documentation included in https://github.com/betaflight/betaflight.com

**Comment by demvlad** [comment]
> Would love to see the documentation included in https://github.com/betaflight/betaflight.com
@haslinghuis Sorry, but i do not understand, what can i see at this page?

**Comment by haslinghuis** [comment]
That is the repo for betaflight.com

**Comment by haslinghuis** [comment]
Sure - you can add to existing or make your own page.

**Comment by demvlad** [comment]
> Sure - you can add to existing or make your own page.

Ok! I've cloned the repo. Will study, how that all work.

**Pull request merged**